server/user.js

```diff
@@ -0,0 +1,212 @@
+    {password: userInfo.password}]};
+  database.getOne(db.collection('userInfo'), query, data, function(data) {
+    var result = authenticateCallback(data);
+    callback(result);
+  });
+
+  var authenticateCallback = function(data) {
+    if (!data) {
+      return {
+        error: 'incorrect username or password'
+      };
+    }
+  };
+};
+
+/**
+ * Gets session ID from database. Regenerate if it is expired.
+ * @param {!mongodb.Db} db Session database.
+ * @param {string} username Username of the POST query.
+ * @param {function(!Object)} callback Callback function.
+ */
+user.authenticate = function(db, username, callback) {
+  var cursor = db.collection('session').find({username: username});
+  var result = [];
+  cursor.each(function(err, doc) {
+    assert.equal(err, null, err);
+    if (doc != null) {
+      result.push(doc);
+    } else {
+      var data = authenticateCallback();
+      callback(data);
+    }
+  });
+  var authenticateCallback = function() {
+    var hasValidSession = false;
+    var sessionIndex = -1;
+    if (result.length) {
+      for (var i = 0; i < result.length; i++) {
+        if (new Date().getTime() < result[i].expiration) {
+          sessionIndex = i;
+          hasValidSession = true;
+          break;
+        }
+      }
+    }
+    if (!result.length || !hasValidSession) {
+      // Don't have username or have username but don't have valid session
+      var cookie = {
+        username: username,
+        sessionId: utils.randomString(),
+        expiration: new Date().getTime() + user.cookieExpireTime
+      };
+      db.collection('session').insertOne(cookie);
+    } else {
+      // Have session and update expire date
+      var cookie = result[sessionIndex];
+      var newExpiration = new Date().getTime() + user.cookieExpireTime;
+      db.collection('session').update(cookie,
+        {$set: {expiration: newExpiration}}, {upsert: false});
+      cookie.expiration = newExpiration;
+    }
+    return {
+      cookie: cookie
+    };
+  };
+};
+
+/**
+ * Validates email address.
+ * @param {string} email
+ * @return {boolean}
+ */
+user.validateEmail = function(email) {
+  return utils.validateRegex(email, user.VALID_EMAIL_REGEX);
+};
+
+/**
+ * Validates username, allows letters, numbers, and underscores, and no less
+ * than 6 characters.
+ * @param {string} username
+ * @return {boolean}
+ */
+user.validateUsername = function(username) {
+  return utils.validateRegex(username, user.VALID_USERNAME_REGEX);
+};
+
+/**
+ * Validates password, allows letters, numbers, and underscores, and no less
+ * than 8 characters.
+ * @param {string} password
+ * @return {boolean}
+ */
+user.validatePassword = function(password) {
+  return utils.validateRegex(password, user.VALID_PASSWORD_REGEX);
+};
+
+/**
+ * Authenticates the signed up user.
+ * @param {!mongodb.Db} db Session database.
+ * @param {!Object} userInfo User Information.
+ * @return {!user.Error}
+ */
+user.signupUser = function(db, userInfo) {
```

Comment:
This is private. The code entry is signUp.

BTW if you prefer "signUp" then the names are to be consistent in cases.